### PR TITLE
chore: Support multi-arch containers build

### PR DESCRIPTION
**Pull request by kahirokunn**
I have tried to start several controllers of aws-controllers-k8s on an ARM EC2 on AWS, but I get the following error.

```
exec ./bin/controller: exec format error
```

To determine the binary format in the Dockerfile, we decided to use TARGETARCH instead of go_arch to facilitate multi-platform build.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kahirokunn** [comment]
If I do them would you please incorporate the PR?

**Comment by acornett21** [comment]
@kahirokunn I'm not a maintainer of ACK, I'm just a contributor.

**Comment by kahirokunn** [comment]
@acornett21 I have responded to all the suggestions you have given me. Are they as intended?

File: Dockerfile [hunk 1]
```diff
@@ -15,14 +17,13 @@ ARG service_alias
...
 ARG build_date
-ARG go_arch=amd64
 # The directory within the builder container into which we will copy our
 # service controller code.
 ARG work_dir=/github.com/aws-controllers-k8s/$service_alias-controller
 WORKDIR $work_dir
 ENV GOPROXY=https://proxy.golang.org|direct
 ENV GO111MODULE=on
-ENV GOARCH=$go_arch
+ENV GOARCH=$TARGETARCH
```
Comment (a-hilaly) [inline Dockerfile:26]: If I understand correctly, this is a variable rename. I'm afraid this might break some of the pipelines that might inject `go_arch` instead of `TARGETARCH`. Are you using this script internally and you rely on this specific variable?

File: scripts/build-controller-image.sh [hunk 2]
```diff
@@ -11,6 +11,11 @@ DOCKERFILE=${DOCKERFILE:-"$DOCKERFILE_PATH"}
 LOCAL_MODULES=${LOCAL_MODULES:-"false"}
 BUILD_DATE=$(date +%Y-%m-%dT%H:%M)
 QUIET=${QUIET:-"false"}
+CONTAINER_OS=${CONTAINER_OS:-"$(go env GOOS)"}
+
+if [ "$CONTAINER_OS" == "darwin" ]; then
+  CONTAINER_OS="linux"
+fi
```
Comment (a-hilaly) [inline scripts/build-controller-image.sh:18]: Most of this code is run by `test-infra` repository pipelines. Are you using darwin instances to build images?

**Comment by kahirokunn** [comment]
@a-hilaly Hi! 
I am an AWS and ARM64 lover.
Could you please review this one as well?

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl [hunk 3]
```diff
@@ -12,12 +12,19 @@ metadata:
     createdAt: {{.CreatedAt}}
     support: {{.Annotations.Support}}
     certified: {{.Annotations.IsCertified}}
+  labels:
+    operatorframework.io/arch.s390x: supported
+    operatorframework.io/arch.amd64: supported
+    operatorframework.io/arch.ppc64le: supported
+    operatorframework.io/arch.arm64: supported
+    operatorframework.io/os.linux: supported
+    operatorframework.io/os.zos: supported
```
Comment (a-hilaly) [inline templates/config/manifests/bases/clusterserviceversion.yaml.tpl:21]: Not familiar of operatorframework, why are these needed?

File: scripts/build-controller-image.sh [hunk 4]
```diff
@@ -100,16 +105,17 @@ if [[ "$LOCAL_MODULES" = "true" ]]; then
   DOCKERFILE="${ROOT_DIR}"/Dockerfile.local
 fi
 
-if ! docker build \
+if ! docker buildx build \
   --quiet="${QUIET}" \
   -t "${AWS_SERVICE_DOCKER_IMG}" \
   -f "${DOCKERFILE}" \
+  --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x,${CONTAINER_OS}/${GOARCH} \
```
Comment (a-hilaly) [inline scripts/build-controller-image.sh:112]: Ditto what does `--platform` do here? release multiple images for the specified platforms?

**Review by a-hilaly — COMMENTED** [review]
Thank you @kahirokunn !

I have few comments in line

/ok-to-test

**Comment by a-hilaly** [comment]
> @a-hilaly Hi!
I am an AWS and ARM64 lover.
Could you please review this one as well?

Happy to help bringing more utilities that will help you build ARM images, let's just make sure that it's done in a way that it's not breaking other system

File: scripts/build-controller-image.sh (hunk 2 above)
Comment (kahirokunn) [reply scripts/build-controller-image.sh:18]: I sometimes use Apple's M1 Mac instance to build and test images.
I think there are quite a lot of Apple M1 Mac users out there, so this should be useful.

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (kahirokunn) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl:21]: @a-hilaly I am not sure either, but they asked me to put it on, so I did.
I don't think this means anything. Do you want me to remove it?
https://github.com/aws-controllers-k8s/code-generator/pull/461#pullrequestreview-1594255248

File: scripts/build-controller-image.sh (hunk 4 above)
Comment (kahirokunn) [reply scripts/build-controller-image.sh:112]: > release multiple images for the specified platforms?

Yes, it is.
I am using an ARM64 instance in ManagedNodeGroup, and when I tried to start it there, it mocked up.
This is because I don't have an ARM64 image.

Comment (kahirokunn) [reply scripts/build-controller-image.sh:112]: When you build with this, multiple platforms are described in the manifest file as follows.

└─> docker manifest inspect nginx

```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:9504f3f64a3f16f0eaf9adca3542ff8b2a6880e6abfb13e478cca23f6380080a",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:6967bfc5a04ad3120ada68fe1c89c1bbac92e4b4793e53a5195be314f75ca17b",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v5"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:ab01fca45d6c3ceb8d7b72364eb03a8b47377b5d42cf1433dcf28d365544f1d8",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:85eabf2757cb5b5b84248d7feb019079501dfd8691fc79b8b1d0ff1591a6270b",
         "platform": {
            "architecture": "arm64",
            "os": "linux",
            "variant": "v8"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:0057a745dbd60ad9c187e89dea876130377a2291df90f28e5fdaaad9c335ae01",
         "platform": {
            "architecture": "386",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:9015ea6b47d4d8d86d4bbd1b0f1e568b56cd703acbd655854716ac6b79c52b1d",
         "platform": {
            "architecture": "mips64le",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:8e0cd0665c6ad7abd9e421d65f826af933cbc7647b4752f354f260a74b92acf9",
         "platform": {
            "architecture": "ppc64le",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1778,
         "digest": "sha256:968ca27471c39e881787859f77ca757690f2c81bfe87c83384853e2d4f8861a8",
         "platform": {
            "architecture": "s390x",
            "os": "linux"
         }
      }
   ]
}
```

Comment (kahirokunn) [reply scripts/build-controller-image.sh:112]: I am also working elsewhere to get the world's container images to work on ARM64.
https://github.com/knative/docs/pull/5273

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (kahirokunn) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl:21]: I don't think it makes sense, so I will remove it.

File: Dockerfile (hunk 1 above)
Comment (kahirokunn) [reply Dockerfile:26]: https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/#93cb
The docker buildx will inject the architecture into the TARGETARCH.
This is practically standard and cannot be changed.

**Comment by kahirokunn** [comment]
I have built it with the arm64 platform added and actually run it in an EKS environment and it works very well.

https://quay.io/repository/kahirokunn/s3-controller/manifest/sha256:7cd95f143a79be3640ee0a87f0ef9c857b27bbf279e186e68e128ce39a6159b3

**Comment by kahirokunn** [comment]
@acornett21 If you create a PR for eks-distro-build-tooling, we will also support the additional platform you specify at that time. Thx 🙏 

```
└─> docker manifest inspect public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base-nonroot:2023-08-24-1692903666.2023
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.index.v1+json",
   "manifests": [
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 506,
         "digest": "sha256:4ea92dc0efd4cc177ee9e2b2a38ccf6f95deb470725db2160e6103bb6dacebf0",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 506,
         "digest": "sha256:ae3b415d4e334c88e12cb6f4d5416f68d04b993ec0e55ddce474a793a479be4d",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```

**Comment by kahirokunn** [comment]
@a-hilaly Do you know why the error occurs? 😢 

```
ERROR: image: "aws-controllers-k8s:s3-v1.0.5-1-g5a93f26-dirty" not present locally
make: *** [Makefile:26: kind-test] Error 1
wrapper.sh] [TEST] Test Command exit code: 2
wrapper.sh] [CLEANUP] Cleaning up after Docker in Docker ...
```

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (acornett21) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: I asked for this, since if ACK is going to distribute a multi-arch image, in order for ACK to be available on architectures other then amd64, this enables the operator to be listed on those clusters.

File: scripts/build-controller-image.sh (hunk 4 above)
Comment (acornett21) [reply scripts/build-controller-image.sh]: `buildx` will build multiple architectures, but also more importantly builds a `manifest-list` media type, that references all of the architectures available for a given image. When a `manifest-list` is stored in an OCI container registry, and an oci engine (podman, docker, crio, etc) goes to pull the image, the registry returns the manifest-list, then the client(podman) sees if the architecture that is running the engine is present (ie if you are running podman on an arm64 system), and arm64 is listed the client then will request the `blobs` for the arm64 image.

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (acornett21) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: I see in another comment, that power/s390x aren't supported by EKS, for now we should just have 
```
operatorframework.io/os.linux: supported
operatorframework.io/arch.amd64: supported
operatorframework.io/os.linux: supported
operatorframework.io/arch.arm64: supported
```

**Comment by a-hilaly** [comment]
>@a-hilaly Do you know why the error occurs? 😢
ERROR: image: "aws-controllers-k8s:s3-v1.0.5-1-g5a93f26-dirty" not present locally
make: *** [Makefile:26: kind-test] Error 1
wrapper.sh] [TEST] Test Command exit code: 2
wrapper.sh] [CLEANUP] Cleaning up after Docker in Docker ...


@kahirokunn not very sure but i suspect `docker buildx` is adding a suffix to the image tag, and our systems are not configured predict those prefixes.
IMO we should also want to change this script so that it's building and pushing the images right https://github.com/aws-controllers-k8s/test-infra/blob/main/cd/scripts/release-controller.sh#L139

**Comment by kahirokunn** [comment]
Now I know why.
It seems that docker does not support Multi Platform images.
Therefore, you have to separate docker build from push.

```
docker buildx build --platform linux/arm64,linux/amd64 -t "kahirokunn/grpc-ping-go" . --load
[+] Building 0.0s (0/0) docker-container:strange_pascal
ERROR: docker exporter does not currently support exporting manifest lists
```

**Comment by kahirokunn** [comment]
@a-hilaly https://github.com/aws-controllers-k8s/test-infra/pull/383
You were pushing with test-infra, I created a PR.
I think this is all that is needed to complete the MULTI ARCH support.

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (kahirokunn) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: I have tested it and it works without it, because the container image supports manifest lists!

File: scripts/build-controller-image.sh (hunk 4 above)
Comment (kahirokunn) [reply scripts/build-controller-image.sh]: Thx 🙏

**Comment by kahirokunn** [comment]
@a-hilaly Hello! What do I need to do to advance this PR?

File: Dockerfile (hunk 1 above)
Comment (kahirokunn) [reply Dockerfile:26]: Now we already support test-infra as well.
Also, my understanding is that the internal pipeline is currently only running on amd64, so it will not break.
https://github.com/aws-controllers-k8s/test-infra/pull/383

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (acornett21) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: @kahirokunn This has nothing to do with the container. This is needed for distribution of the operator bundles to OLM once released. These annotations instruct the catalog-operator in a cluster what operators can be installed for what architectures. If we leave off these annotations, then the operator will not be able to be installed on any ARM OLM distribution.

File: Dockerfile [hunk 5]
```diff
@@ -1,10 +1,12 @@
 # Base image to use at runtime
 ARG base_image=public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base-nonroot:2021-12-01-1638322424
 
+ARG TARGETARCH=amd64
```
Comment (a-hilaly) [inline Dockerfile:4]: Can we name this `ARG target_arch=amd64` to keep the style consistant with the other variables?

**Review by a-hilaly — COMMENTED** [review]
Thank you @kahirokunn !! One last thing.. and we can merge it

File: Dockerfile (hunk 5 above)
Comment (kahirokunn) [reply Dockerfile:4]: https://matsuand.github.io/docs.docker.jp.onthefly/engine/reference/builder/#automatic-platform-args-in-the-global-scope
CleanShot 2023-09-27 at 09 22 15@2x

@a-hilaly Thank you!

TARGETARCH and target_arch in the Dockerfile are incompatible.
For example, you cannot use buildx.
Since TARGETARCH is more compatible with many container build tools, why not prioritize feature compatibility over name compatibility here?

Comment (kahirokunn) [reply Dockerfile:4]: @a-hilaly  I'm a lover of docker such as buildx, so I'm making a suggestion. 🙇‍♂️ 
Of course I think it is an important decision to abandon compatibility with buildx. 👍 
Do you want to discard compatibility and make it a Snake Case? 👀

File: templates/config/manifests/bases/clusterserviceversion.yaml.tpl (hunk 3 above)
Comment (kahirokunn) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: @acornett21 
I think it is backward compatible since it is not annotated from the beginning.
I have never used OperatorFramework at all, so I don't know how to test it or anything.
Could you please open a PR for OperatorFramework yourself?
Thank you 🙏

Comment (acornett21) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: @kahirokunn Please do not resolve this, this is needed like mentioned. Please include this in this PR.

Comment (kahirokunn) [reply templates/config/manifests/bases/clusterserviceversion.yaml.tpl]: @a-hilaly
I put it in because it was requested, but does it seem to be a problem?

https://access.redhat.com/documentation/ja-jp/openshift_container_platform/4.5/html/operators/olm-enabling-operator-for-multi-arch_osdk-generating-csvs

File: Dockerfile (hunk 5 above)
Comment (kahirokunn) [reply Dockerfile:4]: @a-hilaly Do you have an opinion here?

Comment (acornett21) [reply Dockerfile:4]: Trying to conform to buildx does not make sense here, since it's not used in the project. Sticking with snack case and the existing arg of`go_arch` introduces the least number of changes in the process.

**Review by a-hilaly — APPROVED** [review]
/lgtm
